packages/utils/src/carequality/cq-monthly-stats/shared.ts

```diff
@@ -0,0 +1,140 @@
+  cqDirectory: any[];
+  endOfPreviousMonth: string;
+  dayIndex: number;
+};
+
+export type GWStats = {
+  nonErroredResponses?: string;
+  totalDocRetrieved?: string;
+  avgResponseTimeMs?: number;
+};
+
+export type GWWithStats = {
+  [gw: string]: GWStats;
+};
+
+export type ImplementerStats = {
+  implementer?: string;
+  gw?: string;
+  avgResponseTimeMs?: number;
+  nonErroredResponses?: string;
+  totalDocRetrieved?: string;
+};
+
+export type CountPerGW = { [key: string]: [number] };
+
+export async function queryResultsTable<TableResults>(
+  tableName: string,
+  endOfPreviousMonth: string,
+  dayIndex: number
+): Promise<TableResults[]> {
+  const dayPlusOne = dayIndex + 1;
+
+  console.log(`Querying for day ${dayPlusOne}...`);
+
+  const query = `
+      SELECT * FROM ${tableName}
+      WHERE created_at > date '${endOfPreviousMonth}' - INTERVAL '${dayPlusOne} days'
+      AND created_at < date '${endOfPreviousMonth}' - INTERVAL '${dayPlusOne - 1} days'
+    `;
+
+  const results = await readOnlyDBPool.query(query, {
+    type: QueryTypes.SELECT,
+  });
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  const resultsData = results.map((result: any) => result.data) as TableResults[];
+
+  console.log(`Results for day ${dayPlusOne}:`, results.length);
+
+  return resultsData;
+}
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export function getDurationsPerGW(results: any[]): GWWithStats {
+  const durationsPerGW: CountPerGW = {};
+  const xcpdGWStats: GWWithStats = {};
+
+  results.forEach(result => {
+    const gwId = result.gateway.oid || result.gateway.homeCommunityId;
+    const duration = result.duration;
+
+    if (duration) {
+      if (!durationsPerGW[gwId]) {
+        durationsPerGW[gwId] = [duration];
+      }
+
+      durationsPerGW[gwId].push(duration);
+    }
+  });
+
+  for (const [gwId, durations] of Object.entries(durationsPerGW)) {
+    const totalDuration = durations.reduce((acc, curr) => acc + curr, 0);
+    const avgDuration = totalDuration / durations.length;
+
+    xcpdGWStats[gwId] = {
+      avgResponseTimeMs: avgDuration,
+    };
+  }
+
+  return xcpdGWStats;
+}
+
+export function associateGWToImplementer(
+  gwStats: GWWithStats,
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  cqDirectory: any[]
+): ImplementerStats[] {
+  const implementerStats: ImplementerStats[] = [];
+
+  for (const [gateway, stats] of Object.entries(gwStats)) {
+    const implementer = findGWImplementer(gateway, stats, cqDirectory);
+
+    if (!implementer) {
+      console.log(`Could not find implementer for gateway ${gateway}`);
+      continue;
+    }
+
+    implementerStats.push({ implementer, gw: gateway, ...stats });
+  }
+
+  return implementerStats;
+}
+
+function findGWImplementer(
+  gateway: string,
+  stats: GWStats,
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  cqDirectory: any[]
```

Comment:
nit: Could add this type to the script to avoid using `any` on the cq directory:
```
import { Organization } from "@metriport/carequality-sdk/models/organization";

export type CQDirectoryEntryData = {
  id: string; // Organization's OID
  name?: string;
  urlXCPD?: string;
  urlDQ?: string;
  urlDR?: string;
  lat?: number;
  lon?: number;
  state?: string;
  data?: Organization;
  point?: string;
  managingOrganization?: string;
  managingOrganizationId?: string;
  gateway: boolean;
  active: boolean;
  lastUpdatedAtCQ: string;
};
```